File: packages/square-mobile-payments/ios/Plugin/Classes/Options/PaymentParameters.swift
import Foundation
import Capacitor

@objc public class PaymentParameters: NSObject {
    let amountMoney: Money
    let paymentAttemptId: String
    let processingMode: String?
    let referenceId: String?
    let note: String?
    let orderId: String?
    let tipMoney: Money?
    let applicationFee: Money?
    let autocomplete: Bool?
    let delayDuration: String?
    let delayAction: String?

    init(_ obj: JSObject) throws {
        guard let amountMoneyObj = obj["amountMoney"] as? JSObject else {
            throw CustomError.amountMoneyMissing
        }
        self.amountMoney = try Money(amountMoneyObj)

        guard let paymentAttemptId = obj["paymentAttemptId"] as? String else {
            throw CustomError.paymentAttemptIdMissing
        }
        self.paymentAttemptId = paymentAttemptId

        self.processingMode = obj["processingMode"] as? String
        self.referenceId = obj["referenceId"] as? String
        self.note = obj["note"] as? String
        self.orderId = obj["orderId"] as? String

        if let tipMoneyObj = obj["tipMoney"] as? JSObject {
            self.tipMoney = try Money(tipMoneyObj)
        } else {
            self.tipMoney = nil
        }

        if let applicationFeeObj = obj["applicationFee"] as? JSObject {
            self.applicationFee = try Money(applicationFeeObj)
        } else {
            self.applicationFee = nil
        }

        self.autocomplete = obj["autocomplete"] as? Bool
        self.delayDuration = obj["delayDuration"] as? String
        self.delayAction = obj["delayAction"] as? String
    }
}

File: packages/square-mobile-payments/android/src/main/java/io/capawesome/capacitorjs/plugins/squaremobilepayments/classes/options/PaymentParameters.java
package io.capawesome.capacitorjs.plugins.squaremobilepayments.classes.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import io.capawesome.capacitorjs.plugins.squaremobilepayments.classes.CustomExceptions;

public class PaymentParameters {

    @NonNull
    private final Money amountMoney;

    @NonNull
    private final String paymentAttemptId;

    @Nullable
    private final String processingMode;

    @Nullable
    private final String referenceId;

    @Nullable
    private final String note;

    @Nullable
    private final String orderId;

    @Nullable
    private final Money tipMoney;

    @Nullable
    private final Money applicationFee;

    @Nullable
    private final Boolean autocomplete;

    @Nullable
    private final String delayDuration;

    @Nullable
    private final String delayAction;

    public PaymentParameters(@NonNull JSObject obj) throws Exception {
        JSObject amountMoneyObj = obj.getJSObject("amountMoney");
        if (amountMoneyObj == null) {
            throw CustomExceptions.AMOUNT_MONEY_MISSING;
        }
        this.amountMoney = new Money(amountMoneyObj);

        String paymentAttemptId = obj.getString("paymentAttemptId");
        if (paymentAttemptId == null) {
            throw CustomExceptions.PAYMENT_ATTEMPT_ID_MISSING;
        }
        this.paymentAttemptId = paymentAttemptId;

        this.processingMode = obj.getString("processingMode");
        this.referenceId = obj.getString("referenceId");
        this.note = obj.getString("note");
        this.orderId = obj.getString("orderId");

        JSObject tipMoneyObj = obj.getJSObject("tipMoney");
        this.tipMoney = tipMoneyObj != null ? new Money(tipMoneyObj) : null;

        JSObject applicationFeeObj = obj.getJSObject("applicationFee");
        this.applicationFee = applicationFeeObj != null ? new Money(applicationFeeObj) : null;

        this.autocomplete = obj.getBool("autocomplete");
        this.delayDuration = obj.getString("delayDuration");
        this.delayAction = obj.getString("delayAction");
    }

    @NonNull
    public Money getAmountMoney() {
        return amountMoney;
    }

    @NonNull
    public String getPaymentAttemptId() {
        return paymentAttemptId;
    }

    @Nullable
    public String getProcessingMode() {
        return processingMode;
    }

    @Nullable
    public String getReferenceId() {
        return referenceId;
    }

    @Nullable
    public String getNote() {
        return note;
    }

    @Nullable
    public String getOrderId() {
        return orderId;
    }

    @Nullable
    public Money getTipMoney() {
        return tipMoney;
    }

    @Nullable
    public Money getApplicationFee() {
        return applicationFee;
    }

    @Nullable
    public Boolean getAutocomplete() {
        return autocomplete;
    }

    @Nullable
    public String getDelayDuration() {
        return delayDuration;
    }

    @Nullable
    public String getDelayAction() {
        return delayAction;
    }
}
